src/spawn/index.ts

```diff
@@ -0,0 +1,300 @@
+            uuid,
+            name: uuid,
+            mainWindowOptions: {
+                ...position,
+                url,
+                frame: options.frame === undefined ? true : options.frame,
+                state,
+                autoShow: true,
+                saveWindowState: false,
+                defaultWidth: size.x,
+                defaultHeight: size.y
+            }
+        };
+        return startApp(fin.Application.create(data));
+    } else {
+        const queryOptions: Dictionary<string|number|boolean> = {
+            ...position as Required<typeof position>,
+            uuid,
+            url,
+            state,
+            defaultWidth: size.x,
+            defaultHeight: size.y,
+            frame: options.frame === undefined ? true : options.frame,
+            realmName: options.realm || '',
+            enableMesh: options.enableMesh || false,
+            runtime: options.runtime || await fin.System.getVersion(),
+            useService: options.useService !== undefined ? options.useService : true,
+            provider: options.provider || 'local',
+            config: options.config ? JSON.stringify(options.config) : ''
+        };
+
+        let hostname = 'http://localhost:';
+
+        // This can be run in a window or node context. We need to grab the port number from either.
+        if (typeof window !== 'undefined') {
+            hostname += location.port;
+        } else {
+            const {getProjectConfig} = await import('../utils/getProjectConfig');
+            hostname += getProjectConfig().PORT;
+        }
+
+        const manifest = `${hostname}/manifest?${
+            Object.keys(queryOptions)
+                .map(key => {
+                    return `${key}=${encodeURIComponent(queryOptions[key].toString())}`;
+                })
+                .join('&')}`;
+
+        return startApp(fin.Application.createFromManifest(manifest));
+    }
+}
+
+async function createChildWindow(data: Omit<WindowData, 'parent'>): Promise<_Window> {
+    const name: string = data.id || `Win-${Math.random().toString().substr(2, 4)}`;
+    const url = getUrl(data);
+    const position = getWindowPosition(data);
+    const size = getWindowSize(data);
+
+    const options:
+        WindowOption = {...position, name, url, frame: data.frame, autoShow: true, saveWindowState: false, defaultWidth: size.x, defaultHeight: size.y};
+    return fin.Window.create(options);
+}
+
+async function startApp(appPromise: Promise<Application>): Promise<Application> {
+    const app = await appPromise;
+    await app.run();
+    return app;
+}
+
+function getUrl(options: WindowData): string {
+    // Create URL from base URL + queryArgs
+    let url = options.url || 'about:blank';
+    const urlQueryParams = options.queryArgs || {};
+    const urlQueryKeys = Object.keys(urlQueryParams);
+
+    // Resolve relative URL's
+    if (url.indexOf('://') === -1 && url !== 'about:blank') {
```

Comment:
This was here when I moved it in.  I questioned it but was not sure of the deep reaching implications of removing it.  Maybe whoever originally wrote it might have some insights.

Reply:
If a project were to specify a relative URL it would be protected against with that check.  It might be more practical once the PR is in and projects are more likely to have to set URLs.